hms-dbmi-cellenics/ui
# [BIOMAGE-1765] - Notify Slack on error

# Description
#### Issue
We want to be able to monitor fatal error ("It's not you, it's us") errors thrown in production

#### Solution
[NextJS](https://nextjs.org/docs/advanced-features/error-handling#handling-client-errors) and [React](https://reactjs.org/docs/error-boundaries.html) recommends the use of React error boundaries to catch client-side errors. 

I use [a library](https://github.com/bvaughn/react-error-boundary) which implements this. The library exposes a wrapper component with an `onError` handler. Inside this handler, I used a Slack bot to post error logs to the #error-logs channel. The error log that is posted consists of : 
- The error message, without the stack trace. I omitted the stack trace because the stack trace does not accurately identify the position of the error, as the error originates from the compiled JS file, instead of the source file. I wanted to show the exact position of the error as we have in, however it turns out this is more complicated than it seems*.
- The component stack
- The Redux store dump

To correctly map the source of the error, we have to:
- Get the coordinate (line and column) of the error
- Get the source map file
- Remap the apparent coordinate to the source coordinate

However, looking at the error generated in production, the sourcemap does not map to source, but a compiled version of source. Even if I try to read the generated source map, it'll still not give the correct coordinate.

[Next.js version 11 comes with native sourcemap support](https://stackoverflow.com/questions/68505618/how-do-i-generate-sourcemaps-for-nextjs). Our Next.JS version is version 9.5.5. Upgrading to version 11 might allow us to have better source map support. However, we have to be careful when upgrading. 

I suggesst we upgrade first before revisiting this issue.

#### To test
One of the error that still persists is the single-sample DE "go-back" error : https://biomage.atlassian.net/browse/BIOMAGE-1764

To reproduce this : 
- Checkout this branch in your local environment
- Comment line 136 in `_app.jsx` so that logging works for development environment
- Create a new project and upload a single sample. Process the project.
- Go to Data Management, launch DE with any options
- Click "Go Back"
- This should throw an error, which will show an error screen in the UI.
- The error will also throw a log in the #error-logs channel in Slack.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1765

#### Link to staging deployment URL (or set N/A)
https://ui-agi-ui654.scp-staging.biomage.net/

#### Links to any PRs or resources related to this PR
N/A

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR

### Manual/unit testing


### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.


### Documentation updates

### Optional
